### ISSUE TITLE
Tempus_DIRK_Combined_FSA_MPI_1 failing in ATDM build

### ISSUE THREAD

**Issue by fryeguy52** [trilinos/Trilinos#3256]
CC: @trilinos/tempus , @rppawlo  (Trilinos Nonlinear Solvers Product Lead), @bartlettroscoe 
 
## Description

As shown in [this query](https://testing-vm.sandia.gov/cdash/queryTests.php?project=Trilinos&date=2018-08-08&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=4&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-hansen-shiller-gnu-opt-serial&field2=testname&compare2=61&value2=Tempus_DIRK_Combined_FSA_MPI_1&field3=groupname&compare3=61&value3=ATDM&field4=buildstarttime&compare4=83&value4=4%20weeks%20ago) the test:

`Tempus_DIRK_Combined_FSA_MPI_1`

is failing in the build:

`Trilinos-atdm-hansen-shiller-gnu-opt-serial`

prior to 8/3/2018 this test was taking about 6 minutes to complete.  Since then it has completed only once in under 10 minutes but has otherwise failed due to timeout

## Steps to Reproduce

One should be able to reproduce this failure on the machine `hansen` as described in:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md

More specifically, the commands given for the system `hansen` are provided at:

* https://github.com/trilinos/Trilinos/blob/develop/cmake/std/atdm/README.md#shillerhansen

The exact commands to reproduce this issue should be:

```
$ cd <some_build_dir>/

$ source $TRILINOS_DIR/cmake/std/atdm/load-env.sh gnu-opt-serial

$ cmake \
  -GNinja \
  -DTrilinos_CONFIGURE_OPTIONS_FILE:STRING=cmake/std/atdm/ATDMDevEnv.cmake \
  -DTrilinos_ENABLE_TESTS=ON -DTrilinos_ENABLE_Tempus=ON \
  $TRILINOS_DIR

$ make NP=16

$ srun ctest -j16

```

**Comment by ccober6** [comment]
Adding @etphipp as this is his test, and probably the best to know how to reduce the time for this test.

**Comment by etphipp** [comment]
Hmm, when I run the test on my machine, it takes about 30 seconds.  I'm building on shiller now...

**Comment by ccober6** [comment]
Yeah.  When I run these tests on my Mac they take no more than 10s of seconds, but on some platforms they take much longer (>5-10minutes).  Some of this is due to debug mode, but that is not the case here.  I would have hoped that the testing platforms are faster than my 3-year-old Mac. :)

**Comment by bartlettroscoe** [comment]
As shown [here](https://testing-vm.sandia.gov/cdash/testSummary.php?project=1&name=Tempus_DIRK_Combined_FSA_MPI_1&date=2018-08-06) there are many Trilnos builds where this test takes over 300 sec to run and this test also timed out at 10 minutes in the build `Linux-GCC-4.9.3-openmpi-1.8.7_Debug_DEV_Werror` so it is not just this build on this platform that shows this test running slow.  And I don't think those are all debug builds.  And not that some of the PR builds show that test has a runtime of over 300 sec so there is that.

If you want, we can just disable it in this build if you like as described in "Temporarily disable the failing code or test" at:

* https://snl-wiki.sandia.gov/display/CoodinatedDevOpsATDM/Triaging+and+addressing+ATDM+Trilinos+Failures#TriagingandaddressingATDMTrilinosFailures-5.Makesuretheissueisaddressedinatimelyway:

**Comment by etphipp** [comment]
I'm not entirely sure what is going on.  So far I have not gotten the test to take more than 6 minutes to run on hansen.  There is also a massive difference between running "ctest -j16" versus "ctest -j1".  For example, with -j16, this test takes:
```
36/36 Test #20: Tempus_DIRK_Combined_FSA_MPI_1 ...................   Passed  388.51 sec
```
But with -j1 it takes:
```
20/36 Test #20: Tempus_DIRK_Combined_FSA_MPI_1 ...................   Passed   24.13 sec
```
I did notice you have 
```
ATDM_CONFIG_MPI_POST_FLAG="-map-by;socket:PE=16;--oversubscribe"
``` 
for this machine, with I think means that all 16 tests when you run with -j16 will be bound to the same socket and not use the other socket.  I tried changing that to
```
export ATDM_CONFIG_MPI_POST_FLAG="--bind-to;none"
```
but that didn't seem to help.

**Comment by etphipp** [comment]
MPI is also amazingly slow on this machine.  The Tempus version test takes 7 seconds, which is all MPI startup!

**Comment by etphipp** [comment]
I don't understand why it runs an order of magnitude slower with -j16, but probably the thing to do would be to split it up into multiple tests, since it actually tests 10 different time integration methods.

**Comment by bartlettroscoe** [comment]
@etphipp, any suggestions you might have on getting these tests to run faster for this build would be helpful.  But given that the build times for Trilinos are dominating in most cases we have not been as concerned about the test times as long as each test does not take longer than 10 minutes (we had to pick some number).

These tests taking a long time to run is really more of an issue for the Trilinos PR builds since people are really waiting on those.

**Comment by etphipp** [comment]
I've played around with the tests more, and my hypothesis is the FSA tests are much more sensitive to the process migrating between cores, and therefore likely losing their L1 caches.  If I run 2 FSA tests simultaneously bound to separate cores on the same socket, the time for each is about the same as running just one, about 15 seconds.  However if run those tests simultaneously on the same socket, but don't bind them to any cores at all (and let the OS move them around the socket), the time increases by nearly an order of magnitude.  This is true for the non-FSA tests as well, but not to the same degree.

I guess there is still now way to do "ctest -jN" with N>1 and have those N tests bound to separate cores?

**Comment by bartlettroscoe** [comment]
@etphipp said:

> I guess there is still now way to do "ctest -jN" with N>1 and have those N tests bound to separate cores?

Not really.  That is what #2422 is all about.  Kitware started adding some features to CTest for this but a lot more will be needed before that is effective.  This will be a major focus of our FY19 contract with Ktiware and the ATDM Tools & Dev Env project.

**Comment by etphipp** [comment]
I guess the only thing I can do then is to try splitting the test apart into 10 separate tests.  Each test will take about 1/10th the time it does now, so the total time shouldn't change, but maybe if the tests run shorter they will be less likely to migrate.

**Comment by etphipp** [comment]
Splitting the tests apart does seem to help.  When running multiple tests in parallel on hansen, they still run a good 50x slower than they should, but each one is short enough now that it should finish well under 10 minutes.  I'll post a PR soon that splits the DIRK Combined and Staggered FSA tests, which should resolve the issue with this specific test.  Are there other Tempus tests that some times time-out on hansen or other machines?

**Comment by etphipp** [comment]
Thanks @bartlettroscoe.  For now we will just do it this way for this test.  My testing on hansen suggests this is likely to resolve the timeouts.

I wouldn't take my explanation as gospel, it's just a hypothesis.  I don't know what else could be causing it though, because binding the threads to cores fixes it.

**Comment by bartlettroscoe** [comment]
> I wouldn't take my explanation as gospel, it's just a hypothesis

@etphipp, I used the word "postulates" to describe your analysis but otherwise it is hard to explain some of this behavior :-)

But note that while running with higher `ctest -jN` causes significant slowdowns in some tests, like in this case, some level of `N > 1` does result  a lower overall wall-clock time which is most important for getting these builds done given limited computer resources and 24 hour days.  But I would not say we have found the optimal value for `N` for all of these various builds and platforms.

Do you commonly run with `ctest -jN` with higher `N` on your development machines?  Do you not see behavior like this?  I know that on the Linux machines where I run, I don't see this type of massive slowdown when I use `ctest -j16` for example on my Linux machine.

**Comment by etphipp** [comment]
I have a 32 core linux machine where I run with "ctest -j32" all of the time.  I don't see this issue on that machine.

**Comment by etphipp** [comment]
PR #3261 has been merged, so hopefully this will resolve the timeouts.

**Comment by etphipp** [comment]
Can this be close?

**Comment by bartlettroscoe** [comment]
> Can this be closed?

@etphipp, we just need to point to evidence on CDash and then we can close.  The old test is gone and new tests are running it is place, right?

**Comment by etphipp** [comment]
Correct.

**Comment by bartlettroscoe** [comment]
@etphipp, what the new full test names we should be looking for on CDash?

**Comment by etphipp** [comment]
Two tests were replaced by 10 new tests.  The new tests are called "Tempus_DIRK_Combined_FSA_*_MPI_1" and "Tempus_DIRK_Staggered_FSA_*_MPI_1" where * is a different string for each test.

**Comment by bartlettroscoe** [comment]
Okay, we can see the following tests running and passing in the build `Trilinos-atdm-hansen-shiller-gnu-opt-serial` yesterday in [this query](https://testing-vm.sandia.gov/cdash/queryTests.php?project=Trilinos&filtercombine=and&filtercombine=and&date=2018-08-19&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercombine=and&filtercount=3&showfilters=1&filtercombine=and&field1=buildname&compare1=65&value1=Trilinos-atdm-hansen-shiller-gnu-opt-serial&field2=testname&compare2=65&value2=Tempus_DIRK_Combined_&field3=testname&compare3=63&value3=_FSA_) which shows the 10 tests:

| Test Name | Status | Time | Details |
| :-- | :-- | :-- | :-- | 
| Tempus_DIRK_Combined_FSA_Backward_Euler_MPI_1 | Passed | 57s 520ms | Completed |
| Tempus_DIRK_Combined_FSA_EDIRK_2_Stage_3rd_Order_MPI_1 | Passed | 1m 10s 520ms | Completed |
| Tempus_DIRK_Combined_FSA_EDIRK_2_Stage_Theta_Method_MPI_1 | Passed | 1m 10s 260ms | Completed |
| Tempus_DIRK_Combined_FSA_IRK_1_Stage_Theta_MPI_1 | Passed | 1m 2s 210ms | Completed |
| Tempus_DIRK_Combined_FSA_SDIRK_1_Stage_1st_Order_MPI_1 | Passed | 1m 3s 470ms | Completed |
| Tempus_DIRK_Combined_FSA_SDIRK_2_Stage_2nd_Order_MPI_1 | Passed | 2m 15s 310ms | Completed |
| Tempus_DIRK_Combined_FSA_SDIRK_2_Stage_3rd_Order_MPI_1 | Passed | 2m 17s 380ms | Completed |
| Tempus_DIRK_Combined_FSA_SDIRK_3_Stage_4th_Order_MPI_1 | Passed | 3m 5s 460ms | Completed |
| Tempus_DIRK_Combined_FSA_SDIRK_5_Stage_4th_Order_MPI_1 | Passed | 4m 12s 620ms | Completed |
| Tempus_DIRK_Combined_FSA_SDIRK_5_Stage_5th_Order_MPI_1 | Passed | 4m 10s 710ms | Completed |

Some of those are still pretty expensive but all are under 5 minutes.  Therefore, we can close this issue, right?

Closing as complete.

@etphipp, if this is not the right set of tests or you see some other unresolved problem with this, please repoen the issue.

**Comment by etphipp** [comment]
Those are the right tests.

**Comment by bartlettroscoe** [comment]
@etphipp said:

> Those are the right tests.

Good then we are done.  Thanks!
